lib/tasks/marks.rake

```diff
@@ -15,198 +15,275 @@ namespace :db do
         **base_attributes
-      )
-
-      submission_file = new_submission.submission_files.find_by(filename: 'hello.py')
-      base_attributes = {
-        submission_file_id: submission_file.id,
-        is_remark: new_submission.has_remark?,
-        annotation_text_id: AnnotationText.all
-                                          .joins(:annotation_category)
-                                          .where('annotation_categories.assignment': grouping.assignment,
-                                                 'annotation_texts.deduction': nil)
-                                          .where.not('annotation_texts.annotation_category_id': nil)
-                                          .pluck(:id).sample,
-        annotation_number: new_submission.annotations.count + 1,
-        creator_id: Admin.first.id,
-        creator_type: 'Admin',
-        result_id: new_submission.current_result.id
       }
-      TextAnnotation.create(
+
+      base_attributes[:submission_file_id] = submission_file_ids[submission.id.to_s + 'hello.py']
+      base_attributes[:annotation_text_id] = annotation_text_ids[grouping.assignment.id].sample
+      base_attributes[:annotation_number] = 4
+      annotations << {
+        type: 'TextAnnotation',
         line_start: 7,
         line_end: 9,
         column_start: 6,
         column_end: 15,
+        x1: nil,
+        y1: nil,
+        x2: nil,
+        y2: nil,
+        page: nil,
         **base_attributes
-      )
+      }
 
+      total_mark = 0
+      categories_for_assignment = categories_with_criteria[grouping.assignment.id]
+      deductive_category = categories_for_assignment[0]
+      overridden_category = categories_for_assignment[1]
       #Automate marks for assignment using appropriate criteria
-      grouping.assignment.criteria.includes(:marks).each do |criterion|
-        if criterion.class == RubricCriterion
-          random_mark = criterion.max_mark / 4 * rand(0..4)
-        elsif criterion.class == FlexibleCriterion
-          random_mark = rand(0..criterion.max_mark.to_i)
+      criteria[grouping.assignment.id].each do |criterion|
+        override = false
+        if criterion[:type] == 'RubricCriterion'
+          random_mark = criterion[:max_mark] / 4 * rand(0..4)
+        elsif criterion[:type] == 'FlexibleCriterion'
+          if deductive_category[:category_crit_id] == criterion[:id]
+            base_attributes[:annotation_text_id] = deductive_category[:deductive_text]
+            base_attributes[:annotation_number] = 5
+            annotations << {
+              type: 'TextAnnotation',
+              line_start: 4,
+              line_end: 5,
+              column_start: 6,
+              column_end: 15,
+              x1: nil,
+              y1: nil,
+              x2: nil,
+              y2: nil,
+              page: nil,
+              **base_attributes
+            }
+            random_mark = 0
+          elsif overridden_category[:category_crit_id] == criterion[:id]
```

Comment:
@david-yz-liu This line has not failed the last 5 times the task has run, but maybe you can spot flakiness?